tracer/src/Datadog.Trace/Telemetry/Collectors/MetricsTelemetryCollector.cs

```diff
@@ -0,0 +1,478 @@
+    {
+        var buffer = Interlocked.Exchange(ref _buffer, _reserveBuffer);
+        var publicApis = buffer.PublicApiCounts;
+        var counts = buffer.Counts;
+        var countsWithTags = buffer.CountsWithTags;
+        var gauges = buffer.Gauges;
+        var gaugesWithTags = buffer.GaugesWithTags;
+        var distributions = buffer.Distributions;
+        var distributionsWithTags = buffer.DistributionsWithTags;
+
+        var metricData = GetMetricData(publicApis, counts, countsWithTags, gauges, gaugesWithTags);
+        var distributionData = GetDistributionData(distributions, distributionsWithTags);
+
+        // prepare the buffer for next time
+        buffer.Clear();
+        Interlocked.Exchange(ref _reserveBuffer, buffer);
+
+        return new(metricData, distributionData);
+    }
+
+    [Conditional("DEBUG")]
+    private static void AssertTags(Count metric, int actualTags)
+        => Debug.Assert(metric.ExpectedTags() == actualTags, $"Expected {metric} to have {metric.ExpectedTags()} tags, but found {actualTags}");
+
+    [Conditional("DEBUG")]
+    private static void AssertTags(Gauge metric, int actualTags)
+        => Debug.Assert(metric.ExpectedTags() == actualTags, $"Expected {metric} to have {metric.ExpectedTags()} tags, but found {actualTags}");
+
+    [Conditional("DEBUG")]
+    private static void AssertTags(Distribution metric, int actualTags)
+        => Debug.Assert(metric.ExpectedTags() == actualTags, $"Expected {metric} to have {metric.ExpectedTags()} tags, but found {actualTags}");
+
+    private static List<MetricData>? GetMetricData(
+        int[] publicApis,
+        int[] counts,
+        ConcurrentDictionary<MetricKey, int> countsWithTags,
+        int[] gauges,
+        ConcurrentDictionary<MetricKey, int> gaugesWithTags)
+    {
+        var apiLength = publicApis.Count(x => x > 0);
+        var countsLength = counts.Count(x => x > 0);
+        var countsWithTagsLength = countsWithTags.Count(x => x.Value > 0);
+        var gaugesLength = gauges.Count(x => x > 0);
+        var gaugesWithTagsLength = gaugesWithTags.Count(x => x.Value > 0);
+
+        var totalLength = apiLength + countsLength + countsWithTagsLength + gaugesLength + gaugesWithTagsLength;
+        if (totalLength == 0)
+        {
+            return null;
+        }
+
+        var data = new List<MetricData>(totalLength);
+        var timestamp = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
+
+        if (apiLength > 0)
+        {
+            for (var i = 0; i < publicApis.Length; i++)
+            {
+                var value = publicApis[i];
+                if (value > 0 && ((PublicApiUsage)i).ToStringFast() is { } metricName)
+                {
+                    // TODO: should public api be separate metrics, or single tagged metric?
+                    data.Add(
+                        new MetricData(
+                            metricName,
+                            points: new MetricSeries { new(timestamp, value) },
```

Comment:
I looked at all the code paths, and it seems MetricSeries only ever has one element?

Reply:
_Currently_ yes. This is essentially creating the DTO the backend requires though, which is why it's this format.

One situation where we _may_ want to add more points is if we fail to send the metrics. For other telemetry (config etc) we hang onto the values and aggregate them with the new values for sending in the next batch. Currently, for metrics, we're throwing the values away. We could (and in fact, probably should, otherwise counts will be invalid 🤔) instead decide to hang onto them and aggregate with the next batch, in which case we'd have two data points in here